### PR TITLE
Tabular: Added only_breakpoints parameter to .leaderboard()

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

Enables user to filter leaderboard to the relevant models when choosing between inference speed and accuracy.

Example Code:

```
import pandas as pd
import sklearn.datasets

from autogluon.utils.tabular.ml.utils import generate_train_test_split
from autogluon import TabularPrediction as task

LABEL = 'target'


if __name__ == '__main__':
    cancer_data = sklearn.datasets.load_breast_cancer(return_X_y=False)
    X = pd.DataFrame(cancer_data['data'], columns=cancer_data['feature_names'])
    X[LABEL] = cancer_data[LABEL]

    X_train, X_test, y_train, y_test = generate_train_test_split(X=X.drop(LABEL, axis=1), y=X[LABEL], problem_type='binary', test_size=0.2)

    X = X_train
    X[LABEL] = y_train
    X_test[LABEL] = y_test

    X = X.reset_index(drop=True)
    X_test = X_test.reset_index(drop=True)

    predictor_stack = task.fit(
        time_limits=200,
        train_data=X, label=LABEL,
        hyperparameters={
            'GBM': {'num_boost_round': 100},
            'CAT': {'iterations': 100},
            'KNN': {},
        },
        num_bagging_folds=2, stack_ensemble_levels=1,
        num_bagging_sets=2,
    )

    predictor_stack.leaderboard()
    predictor_stack.leaderboard(X_test)

    predictor_stack.leaderboard(only_breakpoints=True)
    predictor_stack.leaderboard(X_test, only_breakpoints=True)
```

Example Output:

```
                                 model  score_val  pred_time_val   fit_time  pred_time_val_marginal  fit_time_marginal  stack_level
0              weighted_ensemble_k0_l2   0.978022       1.165332   7.604892                0.002507           0.333681            2
1              weighted_ensemble_k0_l1   0.975824       0.534352   4.711088                0.005186           0.331667            1
2        LightGBMClassifier_STACKER_l1   0.975824       1.162825   7.271211                0.055676           1.288956            1
3        CatboostClassifier_STACKER_l0   0.973626       0.034891   4.207038                0.034891           4.207038            0
4        CatboostClassifier_STACKER_l1   0.971429       1.148347  10.240033                0.041198           4.257777            1
5        LightGBMClassifier_STACKER_l0   0.969231       0.050904   1.436202                0.050904           1.436202            0
6  KNeighborsClassifierUnif_STACKER_l0   0.940659       0.494274   0.172384                0.494274           0.172384            0
7  KNeighborsClassifierDist_STACKER_l0   0.936264       0.527079   0.166632                0.527079           0.166632            0
8  KNeighborsClassifierDist_STACKER_l1   0.934066       1.631801   6.077256                0.524652           0.095001            1
9  KNeighborsClassifierUnif_STACKER_l1   0.931868       1.625837   6.067221                0.518688           0.084965            1
                                 model  score_test  score_val  pred_time_test  pred_time_val   fit_time  pred_time_test_marginal  pred_time_val_marginal  fit_time_marginal  stack_level
0              weighted_ensemble_k0_l1    0.964912   0.975824        0.486373       0.534352   4.711088                 0.004835                0.005186           0.331667            1
1        LightGBMClassifier_STACKER_l0    0.956140   0.969231        0.054836       0.050904   1.436202                 0.054836                0.050904           1.436202            0
2        CatboostClassifier_STACKER_l1    0.956140   0.971429        1.013497       1.148347  10.240033                 0.025127                0.041198           4.257777            1
3        LightGBMClassifier_STACKER_l1    0.956140   0.975824        1.044854       1.162825   7.271211                 0.056484                0.055676           1.288956            1
4              weighted_ensemble_k0_l2    0.956140   0.978022        1.048074       1.165332   7.604892                 0.003219                0.002507           0.333681            2
5  KNeighborsClassifierUnif_STACKER_l1    0.947368   0.931868        1.448323       1.625837   6.067221                 0.459953                0.518688           0.084965            1
6        CatboostClassifier_STACKER_l0    0.938596   0.973626        0.024499       0.034891   4.207038                 0.024499                0.034891           4.207038            0
7  KNeighborsClassifierDist_STACKER_l0    0.938596   0.936264        0.451996       0.527079   0.166632                 0.451996                0.527079           0.166632            0
8  KNeighborsClassifierUnif_STACKER_l0    0.929825   0.940659        0.457039       0.494274   0.172384                 0.457039                0.494274           0.172384            0
9  KNeighborsClassifierDist_STACKER_l1    0.929825   0.934066        1.448437       1.631801   6.077256                 0.460067                0.524652           0.095001            1
                           model  score_val  pred_time_val  fit_time  pred_time_val_marginal  fit_time_marginal  stack_level
0        weighted_ensemble_k0_l2   0.978022       1.165332  7.604892                0.002507           0.333681            2
1        weighted_ensemble_k0_l1   0.975824       0.534352  4.711088                0.005186           0.331667            1
2  CatboostClassifier_STACKER_l0   0.973626       0.034891  4.207038                0.034891           4.207038            0
                           model  score_test  score_val  pred_time_test  pred_time_val  fit_time  pred_time_test_marginal  pred_time_val_marginal  fit_time_marginal  stack_level
0        weighted_ensemble_k0_l1    0.964912   0.975824        0.507465       0.534352  4.711088                 0.004592                0.005186           0.331667            1
1  LightGBMClassifier_STACKER_l0    0.956140   0.969231        0.059575       0.050904  1.436202                 0.059575                0.050904           1.436202            0
2  CatboostClassifier_STACKER_l0    0.938596   0.973626        0.024316       0.034891  4.207038                 0.024316                0.034891           4.207038            0
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: autogluon/utils/tabular/ml/learner/abstract_learner.py [hunk 1]
```diff
@@ -470,12 +469,20 @@ def predict_and_submit(self, X_test: DataFrame, save=True, save_proba=False):
...
         else:
             trainer = self.load_trainer()
             leaderboard = trainer.leaderboard()
+        if only_breakpoints:
+            if 'score_test' in leaderboard.columns and 'pred_time_test' in leaderboard.columns:
+                score_col = 'score_test'
+                inference_time_col = 'pred_time_test'
+            else:
+                score_col = 'score_val'
+                inference_time_col = 'pred_time_val'
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/learner/abstract_learner.py]: shouldn't these be the `_full` times?

File: autogluon/utils/tabular/ml/utils.py [hunk 2]
```diff
@@ -120,6 +120,34 @@ def dd_list():
...
+    inference_time_min = None
+    for index, row in leaderboard_unique.iterrows():
+        if row[inference_time_col] is None or row[score_col] is None:
+            pass
+        elif inference_time_min is None:
+            inference_time_min = row[inference_time_col]
+            breakpoints.append(index)
+        elif row[inference_time_col] < inference_time_min:
+            inference_time_min = row[inference_time_col]
+            breakpoints.append(index)
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/utils.py]: minor: could be merged into single elif statement with `or`

File: autogluon/utils/tabular/ml/utils.py [hunk 3]
```diff
@@ -120,6 +120,34 @@ def dd_list():
...
+    """
+    Given a set of models, returns in ranked order from best score to worst score models which satisfy the criteria:
+    1. No other model in the set has both a lower inference time and a better or equal score.
+
+    :param leaderboard: Leaderboard DataFrame of model info containing score_col and inference_time_col
+    :param score_col: Column name in leaderboard of model score values
+    :param inference_time_col: Column name in leaderboard of model inference times
+    :return: Subset of the original leaderboard DataFrame containing only models that are a valid optimal choice at different valuations of score and inference time.
+    """
+    leaderboard = leaderboard.sort_values(by=[score_col, inference_time_col], ascending=[False, True]).reset_index(drop=True)
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/utils.py]: should we just always use this sorting in `leaderboard()` (rather than only in `get_leaderboard_breakpoints`)?

File: autogluon/task/tabular_prediction/predictor.py [hunk 4]
```diff
@@ -178,7 +178,7 @@ def evaluate_predictions(self, y_true, y_pred, silent=False, auxiliary_metrics=F
         return self._learner.evaluate(y_true=y_true, y_pred=y_pred, silent=silent,
                                       auxiliary_metrics=auxiliary_metrics, detailed_report=detailed_report)
 
-    def leaderboard(self, dataset=None, silent=False):
+    def leaderboard(self, dataset=None, only_breakpoints=False, silent=False):
```
Comment (jwmueller) [inline autogluon/task/tabular_prediction/predictor.py]: consider clearer argument name instead of `only_breakpoints`, e.g. `only_top_models`.  May also consider mentioning somewhere that this only returns "models on the Pareto frontier of the accuracy/latency tradeoff"

File: autogluon/utils/tabular/ml/learner/abstract_learner.py (hunk 1 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/learner/abstract_learner.py]: These were renamed in the PR that introduced them:

pred_time_val_full -> pred_time_val
pred_time_val -> pred_time_val_marginal

File: autogluon/utils/tabular/ml/utils.py (hunk 3 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/utils.py]: We do, however this is sorted because it has to be sorted for the function to work properly, and we shouldn't trust the input to be sorted.

File: autogluon/task/tabular_prediction/predictor.py (hunk 4 above)
Comment (Innixma) [reply autogluon/task/tabular_prediction/predictor.py]: I didn't know the term Pareto frontier, but that describes it well! Agreed on only_top_models, the tricky part is what the definition of 'top model' is to a user, they might be confused if the 2nd highest acc model isn't present in that case.

**Comment by szha** [comment]
Job PR-397-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-397/1/index.html

File: autogluon/utils/tabular/ml/utils.py (hunk 2 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/utils.py]: Added

File: autogluon/task/tabular_prediction/predictor.py (hunk 4 above)
Comment (Innixma) [reply autogluon/task/tabular_prediction/predictor.py]: Renamed to only_pareto_frontier, added to documentation

**Comment by szha** [comment]
Job PR-397-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-397/2/index.html

**Comment by szha** [comment]
Job PR-397-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-397/3/index.html
